File: src/main/java/jv/collection/stacks/StackArrayResizing.java
package jv.collection.stacks;

import java.util.Iterator;

@SuppressWarnings("unchecked")
public class StackArrayResizing<T> implements Stack<T> {
    private int n = 0;
    private T[] data = (T[]) new Object[1];

    @Override
    public void push(T item) {
        if (n == data.length) {
            resize(2 * data.length);
        }
        data[n++] = item;
    }

    private void resize(int capacity) {
        T[] copy = (T[]) new Object[capacity];
        System.arraycopy(data, 0, copy, 0, n);
        data = copy;
    }

    @Override
    public T pop() {
        T item = data[--n];
        data[n] = null; // remove unused reference from array
        if (n > 0 && n == data.length / 4) {
            resize(data.length / 2);
        }
        return item;
    }

    @Override
    public boolean isEmpty() {
        return n == 0;
    }

    @Override
    public Iterator<T> iterator() {
        return new ArrayIterator<>(data, n);
    }
}

File: src/main/scala/sc/collection/stacks/StackArrayResizing.scala
package sc.collection.stacks

class StackArrayResizing[T >: Null](implicit m: Manifest[T]) extends Stack[T] {
  private var n: Int = 0
  private var data: Array[T] = new Array[T](1)

  def push(item: T) {
    if (n == data.length) {
      resize(2 * data.length)
    }
    data(n) = item
    n += 1
  }

  private def resize(capacity: Int) {
    val copy = new Array[T](capacity)
    System.arraycopy(data, 0, copy, 0, n)
    data = copy
  }

  def pop(): T = {
    n -= 1
    val item = data(n)
    data(n) = null
    if (n > 0 && n == data.length / 4) {
      resize(data.length / 2)
    }
    item
  }

  def isEmpty: Boolean = n == 0


}